### DISCUSSION TITLE
Retargetabe Locking and Linking

### DISCUSSION THREAD

**Discussion by innerand** [picolibc/picolibc#310] [Q&A]
Hi, 

could somebody please explain how linking of the retargetable locks works? 

They dummies are strongly defined in picolibc within [lock.c](https://github.com/picolibc/picolibc/blob/89007348ecf8efe9b913b05b01a64de7a0952dc1/newlib/libc/misc/lock.c). How would the linker know that it should use the implementation of the application and not the one from the library (it does not for me)? 

Many thanks and BR, 
Innerand

**Comment by keith-packard** [comment]
To make this work, you want to define all of the retarget symbols in your application so that they're present before picolibc
is used.

For a working example, you can look at the [Zephyr support for picolibc](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/picolibc/libc-hooks.c). The locking code starts at line 391 under `#ifdef CONFIG_MULTITHREADING`. If you don't define all of the symbols, then the linker will pull in the picolibc support too, and you'll get errors. In particular, you need to declare `__lock___libc_recursive_mutex` along with the `__retarget_lock_*` functions.

**Reply by innerand** [reply]
Thank you for the quick answer, Keith Packard. 

But I am afraid that's what I did and does not work for me. 
Even if I just copy all form lock.c into my application. 
I still get the multiple definitions error for all the locking functions. 

However, I will just exclude lock.c form the picolibc build.

**Reply by keith-packard** [reply]
That happened to me when I forgot to define the picolibc global lock variable, `__lock___libc_recursive_mutex`.

**Reply by keith-packard** [reply]
You can check which symbols you'll need to avoid pulling in lock.c from picolibc by looking at the symbols defined there, for example:
```
$ nm ./newlib/libc.a.p/libc_misc_lock.c.o
00000000 B __lock___libc_recursive_mutex
00000000 T __retarget_lock_acquire
00000000 T __retarget_lock_acquire_recursive
00000000 T __retarget_lock_close
00000000 T __retarget_lock_close_recursive
00000000 T __retarget_lock_init
00000000 T __retarget_lock_init_recursive
00000000 T __retarget_lock_release
00000000 T __retarget_lock_release_recursive
00000000 T __retarget_lock_try_acquire
00000000 T __retarget_lock_try_acquire_recursive
```

**Reply by AntonioND** [reply]
I've been trying to get this to work in my own project. It's a different case as the one mentioned here because I provide a library (`libnds`) that uses picolibc. Essentially, `libnds` implements functions like `open()` or `gettimeofday()` and all the low level code to make them work. Then, when you create your own binary, you link both `libnds` and `libc`, and the full stack is there.

I need to use `-Wl,--start-group $(LIBS) -Wl,--end-group` to make this work, of course, because of the dependency of each library on the other one. All of this has worked very well... until I decided to implement the `__retarget_lock` functions.

All other functions, like `open()`, aren't actually present in picolibc, so the linker looks for them in all other libraries, and it finds them in `libnds`. However, the `__retarget_lock` functions are present in both, so it uses the ones from picolibc.

Forking picolibc, turning them into weak symbols, and linking both libraries doesn't work. I need to use `--gc-sections` to save memory, `libnds` needs to be linked before `libc`. `libnds` doesn't actually use the `__retarget_lock` functions, so `--gc-sections` tells the linker to remove them, and finally the linker sees that `libc` requires them, it finds them in `libc`, so it adds them from there.

So far, the only solution I can find is to remove them completely from picolibc (I'm having problems linking them in C++ projects, but this makes C projects work, and I assume that the linking problem is something unrelated).

**Comment by keith-packard** [comment]
The only way I've used this is to place the locking functions in an object file and insert that into the linker command line before the C library is linked.

If you want to stick them into a library, then you could use `-Wl,-u,__retarget_lock_acquire` early in the linker command line to create an artificial reference to one of the symbols in that file. If your library was the listed before the C library (I think this would work even inside of a `--start-group` `--end-group` pair), then your file would be included first. I think that would even work with `-ffunction-sections` and `-gc-sections` as that process happens after all of the symbols are resolved.

**Comment by keith-packard** [comment]
oh, you could probably just add a reference to one of the symbols from a file in libnds that *will* be referenced from the application.
The ARM crt code does this to make sure an interrupt vector is loaded with this line:
```
	/* Generate a reference to __interrupt_vector so we get one loaded */
	__asm__(".equ __my_interrupt_vector, __interrupt_vector");
```
Replace `interrupt_vector` with any of the retarget lock file and that'll get pulled in before libc is loaded.

**Comment by AntonioND** [comment]
Thanks for your replies!

Both of your options work! I've tried them, and now I'm trying to decide which of the options is cleaner in my codebase.

Also, it looks like my C++ linking issue disappears with both of your suggestions!

Note: Just for the record, in case someone thinks of doing this. A friend suggested me to use `__attribute__((retain))`, but that only works when linking elf files, not with archive files (same issue as weak functions, I guess). I got a lot of `warning: 'retain' attribute ignored [-Wattributes]`.
```

**Comment by steverpalmer** [comment]
I've just fallen over this, and failed when I tried to follow the ideas above.  My fault I'm sure, but I have got a nasty work-around that seems to work - so far.

For completeness, my context is a C software framework using gcc v12.2.1 on a Pico-W _without_ using the pico-sdk.

On the linker command line, I used `-Wl,--wrap=__retarget_lock_init -Wl,--wrap=__retarget_lock_acquire` etc (for all 8 functions).  I then implement the corresponding `__wrap___retarget_lock_init` etc (again for all 8 function) as I required.  The last step is to define an initial value for the `__libc_recursive_mutex` (renamed to `__lock___libc_recursive_mutex` in the image)

```
extern MyRLock __lock___libc_recursive_mutex;
MyRLock __lock___libc_recursive_mutex = MY_RLOCK_INIT;
```

This is wrong for many reasons - but I think it is working for me.  I'm wondering what will happen should Tinystdio start creating locks, but I've not got there yet.

**Comment by keith-packard** [comment]
You shouldn't need to use --wrap with new symbols. Just place all of the symbols in object files linked ahead of libc and libc will use your versions. You can't use a library as they won't get pulled in ahead of the libc uses.

**Reply by steverpalmer** [reply]
Thanks for the quick response.

I don't explicitly mention libc in my link, which I believe is picked up by the picolibc.specs file.  For example, my link command (without my recent --wraps) was 
```
arm-none-eabi-gcc -o build/main.elf -DPICOLIBC_DOUBLE_PRINTF_SCANF -DUSE_FIBERS=1
 -mcpu=cortex-m0plus -mthumb --specs=picolibc.specs -T src/rp2040/picolibc_RP2040.ld -nostartfiles
 -Wl,--wrap=memset -Wl,--wrap=memcpy
 build/panic_handler.o build/destructable.o [snip of ~20 .o's] build/executive/fiber.o build/main.o
```

I did try moving the --specs to the end of the link line, but it made no difference.  Perhaps I should remove the --specs line and elaborate it specifically for my project?

**Reply by keith-packard** [reply]
And your lock functions are in one of the .o files?

**Comment by steverpalmer** [comment]
Yes.

⁣Cheers, Steve​

**Comment by keith-packard** [comment]
Oh, a bunch of in-tree tests use test/lock-valid.c. You might start with that and see how that works; maybe that will help uncover the trouble you're having?

**Comment by steverpalmer** [comment]
Thanks Keith - now it is working!  Must have been finger trouble before - sorry.  For completeness, my `retarget_lock.c` is compiled to a `retarget_lock.o` and included in the list of files passed to the linker.  The link line uses the `--specs=picolibc.specs`, but does not use the `-Wl,--wrap=__retarget_lock_init` I tried previously.  As Keith suggested, if I was doing this again, I'd start from the `test/lock-valid.c`, get that to build and then gradually morph it with your own lock.

**Reply by keith-packard** [reply]
Yeah, you have to define *every* symbol in that file for it to work.

**Reply by steverpalmer** [reply]
Small point, but locking.md mentions `__libc_recursive_mutex`, not `__lock___libc_recursive_mutex`.  Specifically:

    extern struct __lock __libc_recursive_mutex;
    This is the single global lock used for most libc locking. It must be defined in the locking implementation
    in such a way as to not require any runtime initialization.
